src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -594,27 +595,46 @@ internal static bool Dispatch()
             // Before dequeuing the first work item, acknowledge that the thread request has been satisfied
             workQueue.MarkThreadRequestSatisfied();
 
-            object? workItem;
+            object? workItem = null;
             {
-                bool missedSteal = false;
-                workItem = workQueue.Dequeue(tl, ref missedSteal);
+#pragma warning disable CS0162 // Unreachable code detected. SupportsTimeSensitiveWorkItems may be a constant in some runtimes.
+                if (ThreadPool.SupportsTimeSensitiveWorkItems)
+                {
+                    // Alternate between checking for time-sensitive work or other work first, that way both sets of work items
+                    // get a chance to run in situations where worker threads are starved and work items that run also take over
+                    // the thread, sustaining starvation. For example, if time-sensitive work is always checked last here, timer
+                    // callbacks may not run when worker threads are continually starved.
+                    bool dispatchTimeSensitiveWorkFirst = workQueue._dispatchTimeSensitiveWorkFirst;
+                    workQueue._dispatchTimeSensitiveWorkFirst = !dispatchTimeSensitiveWorkFirst;
+                    if (dispatchTimeSensitiveWorkFirst)
+                    {
+                        workItem = workQueue.TryDequeueTimeSensitiveWorkItem();
```

Comment:
Is there any measurable perf impact to checking another queue first before the main queue(s) half the time?

Reply:
I didn't see any noticeable perf difference in tests that utilize this path probably more than in any realistic scenario